### PR TITLE
Make Stdio handle UnwindSafe

### LINKED ISSUE #51863: stdio handles are not `UnwindSafe`
`io::{Stdout, Stderr, StdoutLock, StderrLock}` are not `UnwindSafe`:

```rust
use std::io::{Stdout, Stderr, StdoutLock, StderrLock};
use std::panic::UnwindSafe;

fn assert_unwind_safe<T: UnwindSafe>() {}

fn main() {
    assert_unwind_safe::<Stdout>();
    assert_unwind_safe::<Stderr>();
    assert_unwind_safe::<StdoutLock<'static>>();
    assert_unwind_safe::<StderrLock<'static>>();
}
```

However, because they are protected by mutexes, it doesn't make sense why these handles shouldn't be `UnwindSafe`. It seems to be an oversight because they use `ReentrantMutex` internally, which transitively opts-out of `UnwindSafe` because it contains an `UnsafeCell`.

This makes `io::Stdout` a pain to use with `slog` which requires `Send + Sync + UnwindSafe` on its output, necessitating a second mutex wrapping the handle or a custom wrapper implementing `UnwindSafe`. `AssertUnwindSafe` is not a solution here because it doesn't implement `Write` for its contained type.

Either `ReentrantMutex` should manually implement `UnwindSafe` as well, or the stdio handles should. I also recommend adding delegated impls of `Read`, `Write` and `BufRead` to `AssertUnwindSafe`.

Addendum: I'm still having problems with `slog` but for another reason; `slog_json::Json` is also not `UnwindSafe` or `Sync` because it uses `RefCell` internally.

Addendum 2: because it uses `Mutex` and not `ReentrantMutex`, `Stdin` is already `UnwindSafe + RefUnwindSafe`, and thus `StdinLock` is too. My original code example didn't error for those types but I didn't notice.

### PR DESCRIPTION
**Pull request by estk**
Closes  #51863

This is my first compiler PR. Thanks Niko for the mentor help!

r? @nikomatsakis

### CODE REVIEW

File: src/libstd/io/stdio.rs [hunk 1]
```diff
@@ -712,9 +712,29 @@ pub fn _eprint(args: fmt::Arguments) {
...
     use thread;
     use super::*;
 
+    #[test]
+    fn stdout_unwind_safe() {
+        assert_unwind_safe::<Stdout>();
+    }
+    #[test]
+    fn stdoutlock_unwind_safe() {
+        assert_unwind_safe::<StdoutLock<'static>>();
```
Comment (xfix) [inline src/libstd/io/stdio.rs]: I don't think you need to pass in the explicit lifetime here.

**Review by sugar700 — APPROVED** [review]

**Comment by nikomatsakis** [comment]
> This is my first compiler PR. Thanks Niko for the mentor help!

Glad to hear it! I'm not sure that it was me who helped you, though =) maybe @abonander ?

**Comment by estk** [comment]
Whoops, yep for sure it was @abonander. Thanks to you @abonander. 😄

**Review by abonander — COMMENTED** [review]
If this builds it's probably going to infer a single lifetime whereas I would like to somehow notate a requirement that the trait be implemented for all lifetimes. I'll test locally and see if `for<'a> StdoutLock<' a>` works.

**Comment by estk** [comment]
@abonander interesting, not a feature I knew even existed. It compiled locally and I figured it was fine since the lifetime param seemed irrelevant to whether the StdoutLock and hence ReentrantMutex implemented UnwindSafe and RefUnwindSafe.

Wrt your last comment, StdoutLock is a struct so does not appear to be compatible with HRTBs.

Where do we go from here? Drop the lifetime or pin to `'static`?

**Comment by abonander** [comment]
Lifetimes may not be relevant now but I'm sure one could contrive a few ways to break it without breaking the test as it is written.

My use-case involves `'static` (via `lazy_static!()`) so I would selfishly like a regression test specifically covering that case. Probably just add that to what you have and we'll call it good.

**Comment by abonander** [comment]
And if you don't mind, would you squash the later commits into the first one to keep the history clean?

**Comment by abonander** [comment]
Excellent!

**Pull request merged**